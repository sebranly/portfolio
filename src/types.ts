// TODO: have a glossary on the website
// TODO: explain myProjects series
export enum Tag {
  All = 'All',
  ThreeDimensional = '3D',
  AIML = 'AIML',
  Archived = 'archived',
  C = 'C',
  CSS = 'CSS',
  CLI = 'CLI',
  DNS = 'DNS',
  GitHub = 'GitHub',
  GUI = 'GUI',
  HTML = 'HTML',
  i18n = 'i18n',
  JavaScript = 'JavaScript',
  Live = 'live',
  MyProjects = 'myProjects',
  NintendoDS = 'Nintendo DS',
  OpenGL = 'OpenGL',
  Puzzle = 'puzzle',
  Python = 'Python',
  React = 'React',
  SchoolProject = 'school project',
  SDL = 'SDL',
  VideoGame = 'video game',
  Website = 'website',
  PAlib = 'PAlib',
  PHP = 'PHP',
  SQL = 'SQL',
  TailwindCSS = 'TailwindCSS',
  TypeScript = 'TypeScript',
  Windows = 'Windows'
}

export enum Color {
  Black = 'black',
  White = 'white',
  Gray = 'gray',
  Red = 'red',
  Orange = 'orange',
  Amber = 'amber',
  Yellow = 'yellow',
  Green = 'green',
  Blue = 'blue',
  Indigo = 'indigo',
  Purple = 'purple',
  Pink = 'pink',
  Rose = 'rose'
}

export enum Role {
  Algorithm = 'Algorithm',
  Developer = 'Developer',
  Designer = 'Designer',
  ProductOwner = 'Product Owner',
  Tester = 'Tester'
}

export type Contributor = {
  female?: boolean;
  name: string;
  roles: Role[];
};

export type Project = {
  download?: string;
  title: string;
  contributors?: Contributor[];
  github?: string;
  subtitle: string;
  description: string[];
  tags: Tag[];
  video?: string;
  website?: string;
  years: number[];
  textualSnapshot?: string[];
};

export type Language = {
  shortName: string;
  name: string;
};
