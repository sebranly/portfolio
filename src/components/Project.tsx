import * as React from 'react';
import { Contributor, Project as ProjectType, Role, Tag as TagType } from '../types';
import { pluralize } from '../utils';
import { Tag } from './Tag';
import { useTranslation } from 'react-i18next';

export interface ProjectProps {
  project: ProjectType;
}

const Project: React.FC<ProjectProps> = (props) => {
  const { t } = useTranslation();
  const { project } = props;
  const { contributors, description, github, tags, subtitle, title, website, years } = project;
  const allTags = [...years, ...tags];
  const hasContributors = contributors ? contributors.length > 0 : false;
  const contributorText = pluralize(t('projects.general.contributor'), contributors ? contributors.length : 0);

  return (
    <div className="flex flex-col bg-white px-4 py-4 border-2 w-96 mx-4 sm:mx-2 my-2 rounded-lg border-solid border-gray-50">
      <h2 className="font-bold text-xl">{title}</h2>
      <h3 className="text-lg">{subtitle.startsWith('projects') ? t(subtitle) : subtitle}</h3>
      <div className="border-solid border-black border-2 bg-gray-200 p-2 my-2 text-justify">
        {description.map((sentence: string, index: number) => {
          return (
            <div className="mb-2" key={index}>
              {sentence.startsWith('projects') ? t(sentence) : sentence}
            </div>
          );
        })}
      </div>
      {hasContributors && (
        <div className="inline">
          <div className="font-bold inline">
            {contributorText}
            {t('projects.general.colon')}{' '}
          </div>
          {contributors!.map((contributor: Contributor, index: number) => {
            const { name, roles } = contributor;
            const isLastContributor = index === contributors!.length - 1;
            const suffix = isLastContributor ? '' : ', ';
            const stringifiedContributor = `${name} (${roles
              .map((role: Role) => t(`projects.general.roles.${role}`))
              .join(', ')})${suffix}`;

            return (
              <div className="inline" key={name}>
                {stringifiedContributor}
              </div>
            );
          })}
        </div>
      )}
      {github && (
        <div className="inline">
          <div className="font-bold inline">
            {t('projects.general.code')}
            {t('projects.general.colon')}{' '}
          </div>
          <a
            className="underline inline"
            href={github}
            rel="noopener noreferrer"
            title={`GitHub page for ${title}`}
            target="_blank"
          >
            GitHub
          </a>
        </div>
      )}
      {website && (
        <div className="inline">
          <div className="font-bold inline">
            {t('projects.general.website')}
            {t('projects.general.colon')}{' '}
          </div>
          <a
            className="underline inline"
            href={website}
            rel="noopener noreferrer"
            title={`Website for ${title}`}
            target="_blank"
          >
            {t('projects.general.link')}
          </a>
        </div>
      )}
      <div className="grow h-4" />
      <div>
        {allTags.map((tag: TagType | number) => {
          return <Tag key={tag} tag={tag} />;
        })}
      </div>
    </div>
  );
};

export { Project };
