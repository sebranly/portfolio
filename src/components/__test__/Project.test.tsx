import * as React from 'react';
import { render } from '@testing-library/react';
import { ImageExtension, ImageFolder, Project as ProjectType, Role, Tag } from '../../types';
import { Project } from '../Project';

test('Project', () => {
  const project: ProjectType = {
    contributors: [
      {
        name: 'projects.people.anonymousPerson',
        roles: [Role.Developer]
      }
    ],
    description: ['Some description', 'And another sentence'],
    download: 'some download link',
    images: {
      folder: ImageFolder.ConnectFour,
      images: [{ alt: 'some alt', nb: 1, extension: ImageExtension.PNG }]
    },
    github: 'some github link',
    video: 'some video link',
    website: 'some website link',
    textualSnapshot: [
      '+---+---+---+---+',
      '| #   # |       |',
      '+---+   +   +---+',
      '|     # |   |   |',
      '+   +   +   +   +',
      '|   | #   #   # |',
      '+---+---+---+---+'
    ],
    title: 'Some project',
    subtitle: 'Some subtitle',
    tags: [Tag.Archived, Tag.JavaScript],
    years: [2013]
  };

  const createComponent = () => <Project project={project} />;

  const { container } = render(createComponent());
  expect(container.childNodes).toMatchSnapshot();
});
