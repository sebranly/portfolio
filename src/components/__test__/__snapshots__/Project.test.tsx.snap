// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project 1`] = `
NodeList [
  <div
    class="flex flex-col bg-white px-4 py-4 border-2 w-96 mx-4 sm:mx-2 my-2 rounded-lg border-solid border-gray-50"
  >
    <h2
      class="font-bold text-xl"
    >
      Some project
    </h2>
    <h3
      class="text-lg"
    >
      Some subtitle
    </h3>
    <div
      class="border-solid border-black border-2 bg-gray-200 p-2 my-2 text-justify"
    >
      <div
        class="mb-2"
      >
        Some description
      </div>
      <div
        class="mb-2"
      >
        And another sentence
      </div>
    </div>
    <div
      class="grow h-4"
    />
    <div>
      <div
        class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-black hover:bg-gray-700"
      >
        2013
      </div>
      <div
        class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-gray-500 hover:bg-gray-600"
      >
        projects.general.tags.archived
      </div>
      <div
        class="border-2 border-black border-solid cursor-pointer inline-block rounded mr-2 my-1 px-4 py-1 text-white bg-yellow-500 hover:bg-yellow-600"
      >
        JavaScript
      </div>
    </div>
  </div>,
]
`;
